### PR TITLE
Only advertise and handle features that were reported in client capabilities.

### PR DESCRIPTION
**Pull request by christianparpart**
References: #13253
Based on top of: #12790 

NB: This does not apply to text edit feature, as this should always be present in order to get notified about text updates.

### CODE REVIEW

**Review by Marenz — DISMISSED** [review]

File: libsolidity/lsp/HandlerBase.h [hunk 1]
```diff
@@ -37,6 +39,9 @@ class HandlerBase
 {
 public:
 	explicit HandlerBase(LanguageServer& _server): m_server{_server} {}
+	virtual ~HandlerBase() = default;
+
+	virtual void initialize(Json::Value const& /*_clientCapabilities*/, Json::Value& /*_replyCapabilities*/) {}
```
Comment (cameel) [inline libsolidity/lsp/HandlerBase.h]:
```suggestion
	virtual void initialize(Json::Value const& /*_clientCapabilities*/, Json::Value& /*_serverCapabilities*/) {}
```

Comment (cameel) [inline libsolidity/lsp/HandlerBase.h]: This does not look like generic initialization. Looks like hook for modifying the capability list before it's reported to the client:

```suggestion
	virtual void onReportCapabilities(Json::Value const& /*_clientCapabilities*/, Json::Value& /*_replyCapabilities*/) {}
```

Also, is `_replyCapabilities` always empty or pre-initialized? This should be specified in the docstring.

Finally, I think it would make more sense for the function to only return the list of *extra* capabilities the handler provides instead of being able to modify the whole list. Should one handler really be able to remove capabilities added by another handler?

File: libsolidity/lsp/HandlerBase.h [hunk 2]
```diff
@@ -45,9 +50,9 @@ class HandlerBase
 	/// from the JSON-RPC parameters.
 	std::pair<std::string, langutil::LineColumn> extractSourceUnitNameAndLineColumn(Json::Value const& _params) const;
 
-	langutil::CharStreamProvider const& charStreamProvider() const noexcept { return m_server.compilerStack(); }
-	FileRepository& fileRepository() const noexcept { return m_server.fileRepository(); }
-	Transport& client() const noexcept { return m_server.client(); }
+	langutil::CharStreamProvider const& charStreamProvider() const noexcept;
+	FileRepository& fileRepository() const noexcept;
+	Transport& client() const noexcept;
```
Comment (cameel) [inline libsolidity/lsp/HandlerBase.h:62]: Why?

File: libsolidity/lsp/GotoDefinition.cpp [hunk 3]
```diff
@@ -32,6 +33,21 @@ using namespace solidity::langutil;
...
+	{
+		_replyCapabilities["definitionProvider"] = true;
+		m_server.registerHandler("textDocument/definition", *this);
+	}
+
+	if (_clientCapabilities["textDocument"]["implementation"])
+	{
+		_replyCapabilities["implementationProvider"] = true;
+		m_server.registerHandler("textDocument/implementation", *this);
+	}
```
Comment (cameel) [inline libsolidity/lsp/GotoDefinition.cpp]: The spec says this:
> Client capabilities got introduced with version 3.0 of the protocol. They therefore only describe capabilities that got introduced in 3.x or later. Capabilities that existed in the 2.x version of the protocol are still mandatory for clients. Clients cannot opt out of providing them. So even if a client omits the `ClientCapabilities.textDocument.synchronization` it is still required that the client provides text document synchronization (e.g. open, changed and close notifications).

`textDocument/definition` seems to be one of those 2.x capabilities so older clients might omit it even if they support it.

Might be similar with some other capabilities, I checked only those 2.

File: Changelog.md [hunk 4]
```diff
@@ -6,6 +6,7 @@ Language Features:
 Compiler Features:
 * Yul Optimizer: Allow replacing the previously hard-coded cleanup sequence by specifying custom steps after a colon delimiter (``:``) in the sequence string.
 * Language Server: Implements finding all references as well as semantic highlighting of symbols.
+* Language Server: Constrain server feature set based on advertised client capabilities.
```
Comment (cameel) [inline Changelog.md:10]: Does the LSP spec actually say that the server should trim its capabilities based on what the client requests? I did a quick check and did not see anything about it in the section about the initialization request, though maybe it's buried somewhere deeper.

But just from a common sense perspective - why would the server have different capabilities depending on what the client reports? These things seem completely independent and it seems useful for the client to know about all capabilities even if it does not support them (they could be e.g. displayed to the user as debug info; if would be frustrating for me as a user if it changed based on what the client requests).

File: libsolidity/lsp/LanguageServer.h [hunk 5]
```diff
@@ -115,6 +118,11 @@ class LanguageServer
 	enum class State { Started, Initialized, ShutdownRequested, ExitRequested, ExitWithoutShutdown };
 	State m_state = State::Started;
 
+	// Contains the list of features that are only advertised if advertised by the client
+	// during initialization stage.
+	std::array<std::unique_ptr<HandlerBase>, 5> m_availableHandlers;
```
Comment (cameel) [inline libsolidity/lsp/LanguageServer.h]:
```suggestion
	std::array<std::unique_ptr<HandlerBase>, 5> m_onDemandHandlers;
```

File: libsolidity/lsp/LanguageServer.cpp [hunk 6]
```diff
@@ -130,34 +130,75 @@ Json::Value semanticTokensLegend()
...
+	SourceUnit const& ast = m_server.compilerStack().ast(sourceName);
+	charStreamProvider().charStream(sourceName);
+
+	Json::Value data = SemanticTokensBuilder().build(ast, charStreamProvider().charStream(sourceName));
+
+	Json::Value reply = Json::objectValue;
+	reply["data"] = data;
+
+	client().reply(_id, std::move(reply));
+}
```
Comment (cameel) [inline libsolidity/lsp/LanguageServer.cpp]: The other handlers are not much bigger than this. If they're in separate files, I'd put this one in a separate file too.

File: libsolidity/lsp/LanguageServer.cpp [hunk 7]
```diff
@@ -130,34 +130,75 @@ Json::Value semanticTokensLegend()
 	return legend;
 }
 
+class SemanticTokens: public HandlerBase
```
Comment (cameel) [inline libsolidity/lsp/LanguageServer.cpp]:
```suggestion
class SemanticTokensHandler: public HandlerBase
```

File: libsolidity/lsp/LanguageServer.cpp [hunk 8]
```diff
@@ -130,34 +130,75 @@ Json::Value semanticTokensLegend()
...
+{
+public:
+	explicit SemanticTokens(LanguageServer& _languageServer): HandlerBase(_languageServer) {}
```
Comment (cameel) [inline libsolidity/lsp/LanguageServer.cpp]:
```suggestion
	using HandlerBase::HandlerBase;
```

File: libsolidity/lsp/HandlerBase.h (hunk 2 above)
Comment (christianparpart) [reply libsolidity/lsp/HandlerBase.h:62]: If I keep that implementation in the header, then it'll not compile.

File: Changelog.md (hunk 4 above)
Comment (christianparpart) [reply Changelog.md:10]: Generally speaking, this is PR 1 of 2 (to be followed). this one constrains the functionality as advertised by the LSP client. While it certainly would (in theory) not hurt if the server can respond to more than what the client advertises to support it is also useless if the server provides support for, say, `textDocument/definition`, when the LSP client did not advertise support for it anyways.

Long story short, the VScode solidity extension should in the long run use this LSP server to provide solidity LSP features. But for a smoother transition, the author was requesting to explicitly whitelist/blacklist features that it want's solc to have enabled (that's PR 2 of 2, it's not part of this PR). So this PR is generally also setting the grounding building blocks for the follow-up.

File: libsolidity/lsp/HandlerBase.h (hunk 1 above)
Comment (christianparpart) [reply libsolidity/lsp/HandlerBase.h]: done

File: libsolidity/lsp/GotoDefinition.cpp (hunk 3 above)
Comment (christianparpart) [reply libsolidity/lsp/GotoDefinition.cpp]: We're currently at version 3.17. Version 3.1 came out Feb. 2017, I do not think it makes sense to support anything significantly older in this context.

File: libsolidity/lsp/LanguageServer.cpp (hunk 6 above)
Comment (christianparpart) [reply libsolidity/lsp/LanguageServer.cpp]: @cameel Do you want me to do that in exactly this PR? (Because that's a different topic you're addressing here).

File: libsolidity/lsp/GotoDefinition.cpp (hunk 3 above)
Comment (cameel) [reply libsolidity/lsp/GotoDefinition.cpp]: I mean, older clients is just the justification but regardless of the reason, the spec basically says that `textDocument/definition` is always supported, even if the client tries to say it's not.

File: Changelog.md (hunk 4 above)
Comment (cameel) [reply Changelog.md:10]: > Long story short, the VScode solidity extension should in the long run use this LSP server to provide solidity LSP features. But for a smoother transition, the author was requesting to explicitly whitelist/blacklist features that it want's solc to have enabled (that's PR 2 of 2, it's not part of this PR). So this PR is generally also setting the grounding building blocks for the follow-up.

Oh. That's a very important piece of context that I was missing when I was reviewing this for the first time. I only skimmed the description but I now see that this was discussed in #13253.

This should really be in some comment in the code as well. It's not obvious at all and without this justification the whole mechanism seems just unnecessary and weird.

Also, wouldn't [`ServerCapabilities::experimental`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#serverCapabilities) be a better (and really self-explanatory) way to do this? Dynamically enabling/disabling capabilities seems just like round-about way to go at it.

File: libsolidity/lsp/LanguageServer.cpp (hunk 6 above)
Comment (cameel) [reply libsolidity/lsp/LanguageServer.cpp]: Not sure I understand. Why is it a different topic? It's this PR that adds the handler so it could just as well add it as a separate file. It's not modifying an existing handler. Unless I'm missing some reason that prevented you from doing this?

File: libsolidity/lsp/LanguageServer.h [hunk 9]
```diff
@@ -113,6 +116,11 @@ class LanguageServer
 	enum class State { Started, Initialized, ShutdownRequested, ExitRequested, ExitWithoutShutdown };
 	State m_state = State::Started;
 
+	// Contains the list of features that are only advertised if advertised by the client
+	// during initialization stage.
+	std::array<std::unique_ptr<HandlerBase>, 3> m_onDemandHandlers;
```
Comment (cameel) [inline libsolidity/lsp/LanguageServer.h:121]: Now that I know the context, I think that `experimentalHandlers` would actually be a much better name:
```suggestion
	std::array<std::unique_ptr<HandlerBase>, 3> m_experimentalHandlers;
```

Also, the comment should provide this context and say why we're doing this in the first place. It's probably not a typical LSP behavior.

File: libsolidity/lsp/HandlerBase.h (hunk 2 above)
Comment (cameel) [reply libsolidity/lsp/HandlerBase.h:62]: But why? It did compile until now and at a glance the implementation looks unchanged. Am I missing some detail here?

File: libsolidity/lsp/HandlerBase.h [hunk 10]
```diff
@@ -45,9 +57,9 @@ class HandlerBase
...
```
Comment (cameel) [inline libsolidity/lsp/HandlerBase.h:62]: Just noticed that these return non-const references. Should the functions really be `const` in that case?
